server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamLifecycle.java

```diff
@@ -79,12 +79,14 @@ public static boolean isDataStreamsLifecycleOnlyMode(final Settings settings) {
 
     public static final ParseField ENABLED_FIELD = new ParseField("enabled");
     public static final ParseField DATA_RETENTION_FIELD = new ParseField("data_retention");
+    public static final ParseField EFFECTIVE_RETENTION_FIELD = new ParseField("effective_retention");
+    public static final ParseField RETENTION_SOURCE_FIELD = new ParseField("retention_determined_by");
     public static final ParseField DOWNSAMPLING_FIELD = new ParseField("downsampling");
     private static final ParseField ROLLOVER_FIELD = new ParseField("rollover");
 
     public static final ConstructingObjectParser<DataStreamLifecycle, Void> PARSER = new ConstructingObjectParser<>(
         "lifecycle",
-        false,
+        true,
```

Comment:
I'm a little confused by this change. I *think* it's because we don't expect (or allow) the user to pass in `effective_retention` or `retention_determined_by`, but we use this parser for round-trip serialization tests, which need to use that. Is that right? It might be worth a short comment in the code since the ConstructingObjectParser constructor's javadocs recommend against setting it to true (`This should generally be set to true only when parsing responses from external systems, never when parsing requests from users`).

Reply:
Oh it's more than just the round-trip serialization test -- I tried setting this to `false` to see what else was impacted, and a yaml rest test also failed when the xcontent was being deserialized by DataStreamLifecycle.fromXContent().

Reply:
Thank you for raising this topic @masseyke . I spent some more time looking around and I decided to change the approach for the  following reasons:
- Well, javadoc is very specific for the use of `ignoreUnknownFields`:
> @param ignoreUnknownFields Should this parser ignore unknown fields? This should generally be set to true only when parsing responses from external systems, never when parsing requests from users.

- `DataStreamLifecycle` is used for user input and using `ignoreUnknownFields` means that a user could add different fields including `effective_retention` which will be ignored but the user won't know.
- The problem here I think is that `DataStreamLifecycle` is not only used for user input but it's also deserialised in the cluster state as XContent as part of the data stream. 

As you see we do not have this problem with the rollover information, because they are only used when creating a response to the user. So, I decided to follow this approach. I added a flag, that is set only when we are creating a response to the user and it's false otherwise. This will ensure that the new and unexpected fields will only be present in the user responses and not during other serialisations.

Reply:
I requested another review because the approach changed and I would like your input. Thanks again for bringing this up! I think the original approach was not thought through properly.

Reply:
oh damn, the yaml test failures were because it cannot parse the response, not because the cluster state wasn't retrievable.... back to drawing board.

Reply:
And uncovered a mistake on a payload... that's why the yaml was failing, or at least one of the reasons

Reply:
What you did makes sense to me. I like that we no longer ignore unknown fields.